packages/hyperion-autologging/src/ALSurfaceVisibilityPublisher.ts

```diff
@@ -0,0 +1,144 @@
+        if (event.capability?.trackVisibilityThreshold) {
+          const observer = getOrCreateObserver(event.capability.trackVisibilityThreshold);
+          const { element } = event;
+          /**
+           * IntersectionObserver cannot track display:content styles because
+           * these elements don't have their own "box".
+           * So, instead we have to focus on the children of the parent element
+           */
+          if (ALSurfaceUtils.isSurfaceWrapper(element)) {
+            for (let el = element.firstElementChild; el; el = el.nextElementSibling) {
+              observer.observe(el);
+            }
+          } else {
+            observer.observe(element);
+          }
+          activeSurfaces.set(element, event);
+        }
+
+        break;
+      }
+      case 'unmount_component': {
+        if (event.capability?.trackVisibilityThreshold) {
+          const observer = getOrCreateObserver(event.capability.trackVisibilityThreshold);
+          const { element } = event;
+          if (ALSurfaceUtils.isSurfaceWrapper(element)) {
+            for (let el = element.firstElementChild; el; el = el.nextElementSibling) {
+              observer.unobserve(el);
+            }
+          } else {
+            observer.unobserve(element);
+          }
+          activeSurfaces.delete(element);
+        }
+        break;
+      }
+    }
+
+    function getOrCreateObserver(threshold: number): IntersectionObserver {
+      let observer = observers.get(threshold);
+      if (!observer) {
+        observer = new IntersectionObserver(
+          (entries: IntersectionObserverEntry[], _observer: IntersectionObserver) => {
+            /**
+             * Since surface may have many children that we added above, we need to merge
+             * all the entries, however, in most cases we may have only one entry
+             */
+            const visibleSet = new Map<ALSurfaceMutationEventData, IntersectionObserverEntry[]>();
+            for (const entry of entries) {
+              if (entry.isIntersecting) {
+                const element = entry.target;
+                const surfaceEvent = activeSurfaces.get(element) ?? activeSurfaces.get(element.parentElement); // element or its parent, see above for .observe(...)
+                if (surfaceEvent) {
+                  let entries = visibleSet.get(surfaceEvent);
+                  if (!entries) {
+                    entries = [];
+                    visibleSet.set(surfaceEvent, entries);
+                  }
+                  entries.push(entry);
+                }
+              }
+            }
+            for (const [surfaceEvent, entries] of visibleSet) {
+              let entry = entries[0];
+              __DEV__ && assert(entry != null, 'Unexpected situation');
+              if (entries.length > 1) {
+                // Need to merge the entries
+                // ??
+                console.warn("Don't know yet how to merge entries!");
```

Comment:
This will occur if there are multiple sibling entries that  intersect at once?

Reply:
Yes! The logic to build this is a bit complex, and I wanted to sort out the basics first. Anyway we are using the entry value of the first child, and as of now, I don't know if anyone even cares about this value. So, wanted to postpone adding this more to later.